File: fsharp/pr015.fs
// Problem 15
//
// 19 April 2002
//
//
// Starting in the top left corner of a 2x2 grid, there are 6 routes (without
// backtracking) to the bottom right corner.
//
// [p_015]
//
// How many routes are there through a 20x20 grid?
//
// 137846528820

// Imperative solution.
let baseData n = Array.create (n+1) 1L

let bump vec =
  for i = 0 to Array.length vec - 2 do
    vec.[i+1] <- vec.[i] + vec.[i+1]
  done

let routes n =
  let vec = baseData n in
  for _i = 1 to n do
    bump vec
  done
  vec.[n]

// And a functional solution.
let fbase n = List.init (n+1) (fun _ -> 1L)

let rec fbump = function
  | (a::b::r) -> a :: fbump (a+b :: r)
  | (a:int64 list) -> a

let fsolve n =
  let rec loop items =
    seq {
      yield items
      yield! loop (fbump items)
    }
  loop (fbase n)
  |> Seq.nth n
  |> List.rev
  |> List.head

let pr15 () =
  // routes 20
  fsolve 20

printfn "%d" (pr15 ())

File: rust/src/pr015.rs
// Problem 15
//
// 19 April 2002
//
//
// Starting in the top left corner of a 2x2 grid, there are 6 routes (without
// backtracking) to the bottom right corner.
//
// [p_015]
//
// How many routes are there through a 20x20 grid?
//
// 137846528820

use std::iter;

define_problem!(pr015, 15, 137846528820);

static STEPS: usize = 20;

fn pr015() -> usize {
    let mut values: Vec<_> = iter::repeat(1).take(STEPS + 1).collect();

    for _ in 0 .. STEPS {
        bump(&mut values[..]);
    }

    values[STEPS]
}

fn bump(values: &mut [usize]) {
    for i in 0 .. STEPS {
        values[i+1] = values[i+1] + values[i];
    }
}
